### PR TITLE
Remove redundant final modifier from private methods

### PR DESCRIPTION
**Pull request by mches**
Identifies an opportunity to remove the redundant `final` modifier from private methods (i.e. failing test). Fixes the newly failing test. (TDD)

Note the `final` keyword is not redundant on a non-private static method.

Expands the test case for issue #846, increasing the confidence in merged PR #848 and future refactoring.

### CODE REVIEW

File: java/src/test/resources/source/do_not_format_me/UnnecessaryModifier/NestingRecord_Issue846_Post.java
Comment (blacelle) [inline java/src/test/resources/source/do_not_format_me/UnnecessaryModifier/NestingRecord_Issue846_Post.java:2]: Isn't this duplicating the cases covered in `TestUnnecessaryModifierCases`?

I also feel uncomfortable accumulating these many cases in a single `pre-post` case. (e.g. given file is named `NestingRecord`, as it focus on given specific case).

File: java/src/main/java/eu/solven/cleanthat/engine/java/refactorer/mutators/UnnecessaryModifier.java
Comment (blacelle) [inline java/src/main/java/eu/solven/cleanthat/engine/java/refactorer/mutators/UnnecessaryModifier.java:144]: Given previous comment, I feel such an additional case is better tested with a `@CompareInnerClasses` test-case (if possible of course, which I feel is the case here).

File: java/src/main/java/eu/solven/cleanthat/engine/java/refactorer/mutators/UnnecessaryModifier.java
Comment (blacelle) [inline java/src/main/java/eu/solven/cleanthat/engine/java/refactorer/mutators/UnnecessaryModifier.java:145]: Why not removing `final` if the method is static?
Could we remove final if `private || static`?

Comment (mches) [reply java/src/main/java/eu/solven/cleanthat/engine/java/refactorer/mutators/UnnecessaryModifier.java:145]: Though static methods cannot be overridden, a static final method cannot be hidden, so for that reason alone it's not redundant. I'll link a source for that explanation.

**Comment by mches** [comment]
Thank for the thoughtful review. I'll get it cleaned up.

**Comment by blacelle** [comment]
Could you add a small comment in the code (1 line)and a secomment line with
the url?

If you have a good link for the non static case it would also be helpful
(as I was unsure overriding was also unavailable from another nested class).

**Comment by mches** [comment]
> Could you add a small comment in the code (1 line)and a secomment line with the url? 

Done

**Comment by mches** [comment]
Ready for another pass

File: java/src/test/java/eu/solven/cleanthat/engine/java/refactorer/cases/do_not_format_me/TestUnnecessaryModifierCases.java
Comment (blacelle) [inline java/src/test/java/eu/solven/cleanthat/engine/java/refactorer/cases/do_not_format_me/TestUnnecessaryModifierCases.java:233]: I preferably put these comments in the mutators, not in the unittests. No worry, I'll move them later.

**Review by blacelle — DISMISSED** [review]
Thanks a lot

**Comment by mches** [comment]
What do you think of this simplification? It considers the modifier keyword first rather than last. Do you find it easier to reason about?

```java
	@SuppressWarnings({ "PMD.CognitiveComplexity", "PMD.NPathComplexity" })
	@Override
	protected boolean processNotRecursively(NodeAndSymbolSolver<?> nodeAndSymbolSolver) {
		Node node = nodeAndSymbolSolver.getNode();
		if (!(node instanceof Modifier)) {
			return false;
		}
		var modifier = (Modifier) node;

		if (modifier.getParentNode().isEmpty()) {
			return false;
		}

		var parentNode = modifier.getParentNode().get();

		switch (modifier.getKeyword()) {
		case ABSTRACT:
			return isImplicitlyAbstract(parentNode) && removeModifier(modifier);
		case FINAL:
			return isImplicitlyFinal(parentNode) && removeModifier(modifier);
		case PUBLIC:
			return isImplicitlyPublic(parentNode) && removeModifier(modifier);
		case STATIC:
			return isImplicitlyStatic(parentNode) && removeModifier(modifier);
		default:
			return false;
		}
	}

	private boolean isImplicitlyAbstract(Node node) {
		return isInterfaceLike(node)
				|| (node instanceof MethodDeclaration || node instanceof AnnotationMemberDeclaration)
						&& node.getParentNode().filter(this::isInterfaceLike).isPresent();
	}

	private boolean isImplicitlyFinal(Node node) {
		return node instanceof EnumDeclaration || node instanceof RecordDeclaration
				|| node instanceof FieldDeclaration && node.getParentNode().filter(this::isInterfaceLike).isPresent()
				|| node instanceof MethodDeclaration && ((MethodDeclaration) node).isPrivate();
	}

	private boolean isImplicitlyPublic(Node node) {
		return node.getParentNode().filter(this::isInterfaceLike).isPresent();
	}

	private boolean isImplicitlyStatic(Node node) {
		// interfaces and annotations are implicitly static
		return isInterfaceLike(node)
				// enums are implicitly static
				// https://stackoverflow.com/questions/23127926/static-enum-vs-non-static-enum
				|| node instanceof EnumDeclaration
				// records are implicitly static
				// https://github.com/projectlombok/lombok/issues/3140
				|| node instanceof RecordDeclaration
				|| (node instanceof ClassOrInterfaceDeclaration || node instanceof FieldDeclaration)
						&& node.getParentNode().filter(this::isInterfaceLike).isPresent();
	}
```

File: java/src/test/resources/source/do_not_format_me/UnnecessaryModifier/NestingRecord_Issue846_Pre.java
Comment (mches) [inline java/src/test/resources/source/do_not_format_me/UnnecessaryModifier/NestingRecord_Issue846_Pre.java:30]: This example parses, but doesn't compile. Shall we fix that

```suggestion
	public static final record SomeNestedRecord(int i) {
	}
```

File: java/src/test/resources/source/do_not_format_me/UnnecessaryModifier/NestingRecord_Issue846_Post.java
Comment (mches) [inline java/src/test/resources/source/do_not_format_me/UnnecessaryModifier/NestingRecord_Issue846_Post.java:29]:
```suggestion
	public record SomeNestedRecord(int i) {
```

File: java/src/test/resources/source/do_not_format_me/UnnecessaryModifier/NestingRecord_Issue846_Pre.java
Comment (blacelle) [reply java/src/test/resources/source/do_not_format_me/UnnecessaryModifier/NestingRecord_Issue846_Pre.java:30]: Yes

File: java/src/test/resources/source/do_not_format_me/UnnecessaryModifier/NestingRecord_Issue846_Post.java
Comment (blacelle) [reply java/src/test/resources/source/do_not_format_me/UnnecessaryModifier/NestingRecord_Issue846_Post.java:29]: Yes

**Comment by blacelle** [comment]
This looks definitely simpler to read. Could you make this in a separate PR once this one is merged?

**Comment by mches** [comment]
Certainly

File: java/src/test/resources/source/do_not_format_me/UnnecessaryModifier/NestingRecord_Issue846_Post.java
Comment (mches) [reply java/src/test/resources/source/do_not_format_me/UnnecessaryModifier/NestingRecord_Issue846_Post.java]: Done

**Review by blacelle — APPROVED** [review]

**Comment by blacelle** [comment]
(@mches Do you confirm you can merge by yourself?)

**Comment by mches** [comment]
> (@mches Do you confirm you can merge by yourself?)

I see all green checks that it's ready to merge, but I have no write access to merge it myself.

**Pull request merged**
